token-metadata/program/src/utils.rs

```diff
@@ -830,14 +830,16 @@ pub fn assert_token_program_matches_package(token_program_info: &AccountInfo) ->
     Ok(())
 }
 
+pub fn is_correct_account_type(data: &[u8], data_type: Key, data_size: usize) -> bool {
+    (data[0] == data_type as u8 || data[0] == Key::Uninitialized as u8) && (data.len() == data_size)
```

Comment:
I think I see why this function has to pass `data[0] == Key::Uninitialized` as well as the actual correct `Key`.
1.  Looks like in several places `create_or_allocate_account_raw()` is called on various account types.
1. Right after potentially creating the account, `from_account_info()` is called, before the `key` is set a few lines later.

wdyt about having a `from_uninitialized_account_info()` that is used in those specific cases?  Maybe also another trait function: `safe_uninitialized_deserialize()` that **requires** the key is `Key::Uninitialized` but verifies the size info.

This might be overkill but seems like it makes `is_correct_account_type` a stronger check, and then the few places where uninitialized accounts are deserialized are handled in an explicit way.

Reply:
With green field design this does make sense but feels like overkill for the current state of the program since we still check the account length and none of our accounts have the same length. I'm already having to stop myself from refactoring large swaths of `token-metadata` because it's out-of-scope for this PR which needs to get merged so v1.3 can get to mainnet. However, if you think this is important I can add it.

Reply:
Yeah that is fine, its probably not worth it.